### PR TITLE
Make `BuildDispatch` an `Arc` and box early to remove `UV_STACK_SIZE` hack

### PR DESCRIPTION
**Pull request by konstin**
When looking into the stack size problems with debug mode windows, i saw two things:
* Our stack grows deep (>200 frames) when recursively resolving and installing requirements.
* Our early methods (`main`, `run`, etc.) use exceedingly much stack.

For the first problem, i made `BuildDispatch` an `Arc` and spawned the recursive resolution and installation with `tokio::task::spawn`. It has the nice side effect that most async functions become `Send` and it allows us to spawn arbitrary other futures if we want.

For the second problem, i aggressively boxed futures, in the current diff somewhat arbitrarily.

Together, the test suite passes with 1MB of stack on my linux machine. It should allows us remove the `UV_STACK_SIZE` hack on windows.

### CODE REVIEW

File: crates/bench/benches/uv.rs [hunk 1]
```diff
@@ -152,20 +152,20 @@ mod resolver {
...
-            &git,
-            &in_flight,
+            client.clone(),
+            cache,
+            interpreter.clone(),
+            index_locations,
+            flat_index.clone(),
+            index.clone(),
+            git,
+            in_flight,
```
Comment (BurntSushi) [inline crates/bench/benches/uv.rs:162]: [auto-claim when?](https://smallcultfollowing.com/babysteps/blog/2024/06/21/claim-auto-and-otherwise/)

**Review by BurntSushi — COMMENTED** [review]
Nice! I think the main thing that bothers me here is boxing futures everywhere. Not so much because I'm worried about it impacting perf necessarily, but like, for someone else editing code, how do they know when a future should be boxed? Do we have any feedback mechanism other than "stack exceeded" in CI on Windows?

**Comment by konstin** [comment]
Yeah i forgot that in the PR description: If we commit to changing this i will prune down the boxing. Unfortunately i don't think we have a good measuring system. The best approach would be to have less locals in the high level functions, instead group them and then box the group, but otherwise it comes down to checking the type size and boxing the largest future.

**Comment by BurntSushi** [comment]
Yeah that makes sense. What I figured.

@ibraheemdev Do you have thoughts on this change?

**Pull request closed without merging**